src/compiler/checker.ts

```diff
@@ -23876,6 +23879,25 @@ namespace ts {
                 return type;
             }
 
+            function narrowTypeByPrivateIdentifierInInExpression(type: Type, expr: BinaryExpression, assumeTrue: boolean): Type {
+                const target = getReferenceCandidate(expr.right);
+                if (!isMatchingReference(reference, target)) {
+                    return type;
+                }
+
+                const privateId = expr.left;
+                Debug.assertNode(privateId, isPrivateIdentifier);
+                const symbol = lookupSymbolForPrivateIdentifierDeclaration(privateId.escapedText, privateId);

```

Comment:
Shouldn't this be cached on the private identifier? Is there a call that caches? lookupSymbolForPrivateIdentifierDeclaration doesn't, so I guess that there is a wrapper for it that does. This code should call that wrapper instead. (see comment about creating `checkPrivateIdentifier` -- that is probably the right place if it doesn't already exist.)

Reply:
Thinking about this overnight, it's probably enough to create `checkPrivateIdentifier` in this PR, and fix caching later. The problem predates this PR, but it means that programs with lots of private identifers are likely to have performance problems.

Reply:
You were right, the symbol was cached. So not having to re-resolve the symbol when narrowing now.

Right now I am caching the symbol on the `BinaryExpression` parent. The only reason I am doing this is because otherwise 'findAllRefs' doesn't see the symbol. everything else seems to work if I store the symbol based on the `PrivateIdentifier`.  I'd like to address this, I'm thinking there must be a 'set' that PrivateIds need to be added so they are valid places for 'findAllRefs' to search.

Reply:
Should we be concerned that a code path could reach this line of code *before* the symbol is set?

Reply:
I could put in a call to `checkPrivateIdentifierExpression` if the symbol comes back undefined to be safe.

Reply:
For find-all-refs, you're possibly being tripped up by the fact `PrivateIdentifier` isn't handled by `isExpressionNode`: https://github.com/microsoft/TypeScript/blob/40bd336da52a5d23a8dfdebace89db97a9fe2fbb/src/compiler/utilities.ts#L1904

called from here:

https://github.com/microsoft/TypeScript/blob/7c49552cd516da620b903edccc61af54d1872c3b/src/compiler/checker.ts#L40232

For an `Identifier`, we would perform a normal name resolution:

https://github.com/microsoft/TypeScript/blob/7c49552cd516da620b903edccc61af54d1872c3b/src/compiler/checker.ts#L40245

Instead, you're storing the symbol on the parent binary expression and then looking it up at https://github.com/microsoft/TypeScript/blob/7c49552cd516da620b903edccc61af54d1872c3b/src/compiler/checker.ts#L40284, which seems strange. We don't normally store the symbol on `Identifier` references, since they can have multiple meanings, so it seems a bit strange to cache it for PrivateIdentifier (despite @sandersn's [comment](https://github.com/microsoft/TypeScript/pull/44648/files#r713479416)). However, since a PrivateIdentifier can't have anything other than a value meaning currently, we could probably store it on the NodeLinks for the id itself, rather than its parent expression.

Reply:
thanks! Yes it was the `isExpressionNode` that was causing the find-all-refs issue. I had removed the old `PrivateIdentifierInInExpression` syntax kind from that predicate but not put `PrivateIdentifier`s in.

I rather than `PrivateIdentifier` always returning true for `isExpressionNode`, it only returns true if it's in a valid position to be an 'expression'. This means the checker can use `isExpressionNode` for the grammer checks.
I did think about `utilities.ts` exporting a new more explicit `isPrivateIdentifierInValidExpressionPosition` predicate for the checker to use instead but that didn't feel like it added much value.

I am now caching the resolved symbol on the privateId itself, not the parent. As private-identifiers can only reference one member of a syntactically outer class, and not come from a different script/module. This seems safe.

Reply:
> Should we be concerned that a code path could reach this line of code before the symbol is set?

While this wasn't happening in the tests. It did happen via the language server, and wasn't getting narrowed types on hover in vscode. I've added a call to `checkPrivateIdentifierExpression` when the symbol is `undefined` which fixes the issue.

Is there an alternative approach than checking for `undefined`, to see if the type-check has already happened? For the case when getting `undefined` is actually a cache-hit but because of a typo in the privateIdentifier it has no symbol to resolve to - yet `checkPrivateIdentifierExpression` would keep being called.